cmake/Modules/MacroQbtCommonConfig.cmake

```diff
@@ -0,0 +1,99 @@
+# Set common variables and create some interface-only library targets
+# that some or all other targets will link to, either directly or transitively,
+# to consume common compile options/definitions
+
+macro(qbt_common_config)
+
+    # treat CXX_STANDARD as a requirement if user manually specifies CMAKE_CXX_STANDARD
+    set(CMAKE_CXX_STANDARD_REQUIRED ON)
+
+    # these definitions are only needed for calls to
+    # lt::generate_fingerprint and for the qbittorrent.rc file on Windows
+    add_library(qbt_version_definitions INTERFACE)
+
+    target_compile_definitions(qbt_version_definitions
+        INTERFACE
+            QBT_VERSION_MAJOR=${PROJECT_VERSION_MAJOR}
+            QBT_VERSION_MINOR=${PROJECT_VERSION_MINOR}
+            QBT_VERSION_BUGFIX=${PROJECT_VERSION_PATCH}
+            QBT_VERSION_BUILD=${PROJECT_VERSION_TWEAK}
+    )
+
+    set(PROJECT_VERSION "${PROJECT_VERSION_MAJOR}.${PROJECT_VERSION_MINOR}.${PROJECT_VERSION_PATCH}.${PROJECT_VERSION_TWEAK}${QBT_VER_STATUS}")
+
+    add_library(qbt_common_cfg INTERFACE)
+
+    # Full C++ 14 support is required
+    target_compile_features(qbt_common_cfg
+        INTERFACE
+            cxx_std_14
+            cxx_contextual_conversions
+            cxx_binary_literals
+            cxx_decltype_auto
+            cxx_lambda_init_captures
+            cxx_generic_lambdas
+            cxx_variable_templates
+            cxx_relaxed_constexpr
+            cxx_attribute_deprecated
+            cxx_digit_separators
+    )
+
+    target_compile_definitions(qbt_common_cfg
+        INTERFACE
+            QBT_VERSION="v${PROJECT_VERSION}"
+            QBT_VERSION_2="${PROJECT_VERSION}"
+            QT_DEPRECATED_WARNINGS
+            QT_NO_CAST_TO_ASCII
+            QT_NO_CAST_FROM_BYTEARRAY
+            QT_USE_QSTRINGBUILDER
+            QT_STRICT_ITERATORS
+            $<$<NOT:$<CONFIG:Debug>>:QT_NO_DEBUG_OUTPUT>
+            $<$<AND:$<BOOL:${UNIX}>,$<NOT:$<BOOL:${APPLE}>>,$<NOT:$<BOOL:${CYGWIN}>>>:_DEFAULT_SOURCE>
```

Comment:
@Chocobo1, are you okay with this monstrous notation instead of human readable one like the following?
```cmake
if (UNIX AND NOT APPLE AND NOT CYGWIN)
        target_compile_definitions(qbt_common_cfg INTERFACE _DEFAULT_SOURCE)
endif()
```
(in all similar cases)

Reply:
I'm not familiar with the usage of generator expression, maybe FranciscoPombal can provide the reasons of using them.

Reply:
@glassez @Chocobo1 

The reasoning for using generator expressions was outlined in part of this comment (https://github.com/qbittorrent/qBittorrent/pull/12746#issuecomment-657070063):

> While attempting to solve compilation with MinGW I also made everything more target-based with the help of generator expressions (that commit is not squashed, for now). As a side-effect, the platform-specific requirements for each target are much clearer now; where possible, all the information is conveyed in single `target_*` calls, instead of having to navigate through multiple, scattered, possibly nested `if ()` chains that affect targets everywhere. Note: the only exception is one `include_directories()` statement. I think that's the only directory-scope command left, but I'm ok with that for now.

and further expanded upon in this whole comment: https://github.com/qbittorrent/qBittorrent/pull/12746#issuecomment-657476593.

Basically, though the syntax is a bit ugly, I think that is an acceptable price to pay for the benefits this change yields.
To be fair, `$<$<AND:$<BOOL:${UNIX}>,$<NOT:$<BOOL:${APPLE}>>,$<NOT:$<BOOL:${CYGWIN}>>>:_DEFAULT_SOURCE>` is a bit of a worst-case scenario, all things considered. It can't (or at least shouldn't) get worse than that.

Reply:
CMake is my default build system but I never use generator expression, they're are not build for humans c++ is already hard why're you trying to make build system hard too, I'll personally like you to not use them at all.

If() statements may seem redundant but they tend too be much more readable.

Reply:
My opinion is avoid generator expressions whenever possible. I prefer more readability friendly code.
In addition, "overusing generator expressions" is something that fell on our heads after this PR was almost completed, and we were trying to finish it. Well, now here's another controversial thing that could have had its own PR rather than slowing this one down.

Reply:
@glassez @jagannatharjun 

> CMake is my default build system but I never use generator expression, they're are not build for humans c++ is already hard why're you trying to make build system hard too, I'll personally like you to not use them at all.
>
> If() statements may seem redundant but they tend too be much more readable.

> My opinion is avoid generator expressions whenever possible. I prefer more readability friendly code.

Did you even read my previous posts explaining the benefits? You guys are arguing "don't use this nice feature of modern CMake because I don't understand it". Despite the somewhat ugly syntax, generator expressions are the best (and often the only easy correct way, see video below) to do certain things. Generator expressions _is_ more friendly code, once you look past the syntax.

There are many nice videos on YT explaining modern CMake (mostly from C++ conferences from 2017 to present) and generator expressions, including what they are used for and why they're needed and better than the alternative. I encourage you to watch at least this ~10 minute part explaining genexps of this particular talk about modern CMake, https://youtu.be/mn1ZnO3MtVk?t=2055, starting at 34:15 through 45:11 (the link already includes the correct timestamp).

Some highlights:

- 35:44 - "Abstraction targets" - what I attempted to do with the `qbt_common_cfg` and `qbt_version_definitions` targets inside `MacroQbtCommonConfig.cmake`
- 38:49 - Relevant example of generator expressions preventing buggy and non-portable `if ()`-based code
- 43:08 - Relevant example of an imperative, non-modern CMake logic snippet being modernized and simplified with genexps.

> In addition, "overusing generator expressions" is something that fell on our heads after this PR was almost completed, and we were trying to finish it. Well, now here's another controversial thing that could have had its own PR rather than slowing this one down.

The original motivation to also switch to use more generator expressions was to also tackle the MinGW build logic, which I originally intended to not even touch (perhaps leaving that to a later PR). I noticed it was becoming too hard to easily reason about target and platfom requirements with all the `if ()`s and scattered imperative logic, so I turned it all to targets + genexp. It made everything about different multiplatform requirements easier to reason about and conclude that currently, there isn't even any special logic required to build with MinGW. The previous logic was even kind of dangerous, as it included old-style calls to `target_link_libraries()`, if I'm not mistaken.

The whole build system is now easy to understand, change and extend, no matter the platform/OS (okay, probably excluding the translations part in `app`, which is still somewhat messy, but as I said before, there is little we can do about it for now), as long as one takes the time to understand genexps. It's a small investment, and not doing so is like saying "I want to read/write modern C++, but I don't want to learn anything new about C++ >= 11".
I think we are on the same page concerning the desire (and a need actually, considering the upcoming deprecation of qmake) to move to CMake as our build system - we should understand the tools that we use.